op-chain-ops/cmd/celo-migrate/beacon.go

```diff
@@ -0,0 +1,254 @@
+// looks back from there to find the first finalized block.
+func (c *beaconClient) MostRecentFinalizedL1BlockAtTime(l2StartTimeSeconds uint64) (common.Hash, error) {
+	// Find the epoch starting at or before the L2 start time.
+	epochNumber := SlotAtOrBefore(l2StartTimeSeconds) / beaconSlotsPerEpoch
+
+	// This epoch is guaranteed to not be complete at L2 start time (if the L2 start time falls in the last
+	// second of the epoch the epoch is still not complete, and if it was we'd be selecting the next epoch)
+	// The previous epoch is the most recent completed epoch.
+	// The one prior to that is the most recent justified epoch.
+	// And the first block of the justified epoch (the epoch boundary block) will be finalized.
+	// This is assuming that there was at least 2/3 participation in the completed and justified epochs.
+	var epoch *Epoch
+	var err error
+	names := [2]string{"completed", "justified"}
+
+	// Check the most recent completed and justified epochs had a participation rate of at least 0.67.
+	for i := uint64(1); i <= 2; i++ {
+		epoch, err = c.Epoch(context.Background(), epochNumber-i)
+		if err != nil {
+			return common.Hash{}, fmt.Errorf("error fetching epoch %d: %w", epochNumber-i, err)
+		}
+		if epoch.Data.Globalparticipationrate < 0.67 {
+			return common.Hash{}, fmt.Errorf("most recent %s epoch before the L2 start time (%d) has less than 0.67 participation rate (%.2f)", names[i-1], l2StartTimeSeconds, epoch.Data.Globalparticipationrate)
+		}
+	}
+	// Calculate the first slot of the most recent justified epoch.
+	mostRecentFinalizedSlot := (epochNumber - 2) * beaconSlotsPerEpoch
+
+	// Find the most recent actual finalized block, slots can be empty so we
+	// search back if we encounter empty slots. We check up to 10 slots, if they
+	// are all empty something serious is wrong with the L1 so we abort.
+	var beaconBlock *BeaconBlock
+	for i := uint64(0); i < 10; i++ {
+		beaconBlock, err = c.BeaconBlock(context.Background(), mostRecentFinalizedSlot-i)
+		if errors.Is(err, ethereum.NotFound) {
+			// If there is not block for this slot then skip to the next.
+			continue
+		}
+		if err != nil {
+			return common.Hash{}, fmt.Errorf("error fetching beacon block at slot %d: %w", mostRecentFinalizedSlot, err)
+		}
+		if !beaconBlock.Finalized {
+			return common.Hash{}, fmt.Errorf("expecting beacon block at slot %d to be finalized", mostRecentFinalizedSlot)
+		}
+		break // We found a good block.
+	}
+	if beaconBlock == nil {
+		return common.Hash{}, fmt.Errorf("failed to find finalized block searching up to 10 slots back from the most recent finalized slot (%d) at the L2 fork time (%d)", mostRecentFinalizedSlot, l2StartTimeSeconds)
+	}
+	return common.HexToHash(beaconBlock.Data.Message.Body.ExecutionPayload.BlockHash), nil
+}
+
+// Gets the epoch from the beaconcha.in api
+func (c *beaconClient) Epoch(ctx context.Context, num uint64) (epoch *Epoch, err error) {
```

Comment:
`ctx` is not used

Reply:
Thanks, realised for some reason my linter had stopped working, now re-instated